lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -11599,9 +11619,33 @@ void Emit(ParseNode* pnode, ByteCodeGenerator* byteCodeGenerator, FuncInfo* func
         byteCodeGenerator->Writer()->Element(
             Js::OpCode::LdElemI_A, pnode->location, protoLocation, pnode->AsParseNodeBin()->pnode2->location);
 
+
         ENDSTATEMENET_IFTOPLEVEL(isTopLevel, pnode);
         break;
     }
+
+    case knopOptChain: {
```

Comment:
It's not immediately clear what cases send us here as opposed to using the EmitNullPropogation above, is this when there's multiple levels?

Reply:
I wanted to reuse as much of the existing logic as possible.
So instead of parsing a special AST for optional-chaining I wrap the whole expression / chain into a `UniNode` with `knopOptChain`.

https://github.com/chakra-core/ChakraCore/blob/cc547645d85b9e4db2070cc000449266ac9a3167/lib/Parser/Parse.cpp#L4316-L4323

That way every byte-code emission of an optional-chain will go through `knopOptChain` and I can create a single `skipLabel` and control its life-time in `FuncInfo` via the c++ call-stack.

https://github.com/chakra-core/ChakraCore/blob/cc547645d85b9e4db2070cc000449266ac9a3167/lib/Runtime/ByteCode/ByteCodeEmitter.cpp#L11630-L11632

Every `?.` token marks its AST node as `isNullPropagating`.
A call to `EmitNullPropagation` can then check for `null` (or `undefined`) and use the single `skipLabel` from `FuncInfo` to skip the whole `knopOptChain` sub-tree (the whole chain).

Reply:
Please could you add a few comments in the source explaining this?

In particular (assuming I've understood correctly) would be good to have a comment by EmitNullPropogation explaining it will only be called within a knopOptChain.